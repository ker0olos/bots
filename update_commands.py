import os
import sys
import json
import typing
from enum import Enum

import requests

APP_ID = os.getenv("APP_ID")

BOT_TOKEN = os.getenv("BOT_TOKEN")

GUILD_ID = os.getenv("GUILD_ID")

url = (
    f"https://discord.com/api/v10/applications/{APP_ID}/commands"
    if GUILD_ID is None
    else f"https://discord.com/api/v10/applications/{APP_ID}/guilds/{GUILD_ID}/commands"
)


class Type(Enum):
    SUB_COMMAND = 1
    SUB_COMMAND_GROUP = 2
    STRING = 3
    INTEGER = 4  # Includes all channel types + categories
    BOOLEAN = 5
    USER = 6
    CHANNEL = 7  # Includes all channel types + categories
    ROLE = 8
    MENTIONABLE = 9  # Includes users and roles
    NUMBER = 10  # Any double between -2^53 and 2^53
    ATTACHMENT = 11  # attachment object


class Option:
    def __init__(
        self,
        name: str,
        desc: str,
        type: Type,
        required: bool = True,
        # options: typing.List | None = None,
    ):
        # if options is not None:
        #     options = [option.__dict__ for option in options]
        # else:
        #     options = []

        self.name = name
        self.type = type.value
        self.name = name
        self.description = desc
        self.required = required
        # self.options = options


def make_command(
    name: str,
    desc: str | None = None,
    options: typing.List[Option] | None = None,
    aliases: typing.List[str] | None = None,
    canary_only: bool = False,
):
    if (canary_only) and (GUILD_ID is None):
        return []

    if canary_only:
        desc = f"{desc} (canary only)"

    commands = [
        {
            "name": name,
            "description": desc,
            "options": [option.__dict__ for option in options]
            if options is not None
            else [],
        }
    ]

    if aliases is not None:
        for alias in aliases:
            copy = commands[0].copy()
            copy["name"] = alias
            commands.append(copy)

    return commands


def load_manifest(filepath: str):
    with open(filepath + "/manifest.json", "r", encoding="utf-8") as file:
        data = file.read()

    manifest = json.loads(data)

    commands = []

    for name in manifest["commands"]:
        desc = manifest["commands"][name]["description"]
        options = manifest["commands"][name]["options"]

        commands += make_command(
            name,
            desc,
            [
                Option(
                    name=opt["id"],
                    desc=f'{opt["description"]} ({manifest["title"]})',
                    type=Type[opt["type"].upper()],
                )
                for opt in options
            ],
        )

    return commands


def set_commands(commands):
    print(os.getenv("GITHUB_REF_NAME"))

    if GUILD_ID is None:
        print("Updating global commands for production bot\n\n")
    else:
        print("Updating guild commands for canary bot\n\n")

    response = requests.put(
        url, headers={"Authorization": f"Bot {BOT_TOKEN}"}, json=commands, timeout=15000
    ).json()

    print(response)

    if "code" in response:
        sys.exit(1)


if __name__ == "__main__":
    set_commands(
        # standard gacha commands
        # uses characters and media from
        # all builtin, community, and manual repos
        make_command(
            name="anime",
            desc="Search for an anime/manga",
            options=[
                Option(
                    name="query",
                    desc="The title for an anime/manga",
                    type=Type.STRING,
                ),
                Option(
                    name="debug",
                    desc="Display the nerdy stuff",
                    type=Type.BOOLEAN,
                    required=False,
                ),
            ],
            aliases=["manga"],
        )
        + make_command(
            name="character",
            desc="Search for a character",
            options=[
                Option(
                    name="query",
                    desc="The title of the character",
                    type=Type.STRING,
                ),
                Option(
                    name="debug",
                    desc="Display the nerdy stuff",
                    type=Type.BOOLEAN,
                    required=False,
                ),
            ],
            aliases=["debug"],
        )
        + make_command(
            name="songs",
            desc="Search for the OP/ED and theme songs of an anime",
            options=[
                Option(
                    name="query",
                    desc="The title for an anime/manga",
                    type=Type.STRING,
                )
            ],
            aliases=["themes"],
        )
        + make_command(
            name="gacha",
            desc="An experimental/ephemeral gacha command",
            aliases=["w", "pull", "roll"],
            canary_only=True,
        )
        + make_command(
            name="force_pull",
            desc="Force a gacha pull",
            options=[
                Option(
                    name="id",
                    desc="The id of the character",
                    type=Type.STRING,
                )
            ],
            canary_only=True,
        )
        # repo management commands
        + make_command(
            name="repo",
            options=[
                Option(
                    name="builtin",
                    desc="Show all built-in enabled-by-default repositories",
                    type=Type.SUB_COMMAND,
                    required=False,
                )
            ],
            canary_only=True,
        )
        # non-standard (eternal) commands
        # non-gacha commands (specific one-task commands)
        + load_manifest("./repos/anilist")
        + load_manifest("./repos/utils")
    )
