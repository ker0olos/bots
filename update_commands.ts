import { load as Dotenv } from 'https://deno.land/std@0.175.0/dotenv/mod.ts';

import { green } from 'https://deno.land/std@0.175.0/fmt/colors.ts';

import { Manifest } from './src/types.ts';

try {
  await Dotenv({ export: true, allowEmptyValues: true });
} catch {
  //
}

const APP_ID = Deno.env.get('APP_ID');

const BOT_TOKEN = Deno.env.get('BOT_TOKEN');

const GUILD_ID = Deno.env.get('GUILD_ID');

if (!APP_ID) {
  throw new Error('APP_ID is not defined');
}

if (!BOT_TOKEN) {
  throw new Error('BOT_TOKEN is not defined');
}

enum Type {
  'SUB_COMMAND' = 1,
  'SUB_COMMAND_GROUP' = 2,
  'STRING' = 3,
  'INTEGER' = 4, //  Includes all channel types + categories
  'BOOLEAN' = 5,
  'USER' = 6,
  'CHANNEL' = 7, // Includes all channel types + categories
  'ROLE' = 8,
  'MENTIONABLE' = 9, // Includes users and roles
  'NUMBER' = 10, // Any double between -2^53 and 2^53
  'ATTACHMENT' = 11, // attachment object
}

enum Permission {
  'OWNER' = 0,
  'ADMINISTRATORS' = 1 << 3,
  'MANAGE_GUILD' = 1 << 5,
}

const Option = (
  { name, description, type, autocomplete, optional }: {
    type: Type;
    name: string;
    description: string;
    autocomplete?: boolean;
    optional?: boolean;
  },
) => ({
  name,
  description,
  autocomplete,
  type: type.valueOf(),
  required: !optional,
});

const Command = ({
  name,
  description,
  options,
  aliases,
  defaultPermission,
  devOnly,
}: {
  name: string;
  description: string;
  options?: ReturnType<typeof Option>[];
  aliases?: string[];
  defaultPermission?: Permission;
  devOnly?: boolean;
}) => {
  if (devOnly && !GUILD_ID) {
    return [];
  }

  if (devOnly) {
    description = `${description} (Developer Only)`;
  }

  const commands = [{
    name,
    description,
    'default_member_permissions': defaultPermission,
    options,
  }];

  aliases?.forEach((alias) =>
    commands.push({
      ...commands[0],
      name: alias,
    })
  );

  return commands;
};

type CommandsArray = ReturnType<typeof Command>;

const Pack = (path: string): CommandsArray => {
  const data = Deno.readTextFileSync(path + '/manifest.json');
  const manifest: Manifest = JSON.parse(data);

  const commands: CommandsArray = [];

  if (manifest.commands) {
    Object.entries(manifest.commands).forEach(([name, command]) => {
      const options = command.options;

      const description = `${command.description} (${
        manifest.title || manifest.id
      })`;

      commands.push(...Command({
        name,
        description,
        options: options.map((opt) => (Option({
          name: opt.id,
          description: opt.description,
          type: Type[opt.type.toUpperCase() as keyof typeof Type],
          optional: !opt.required,
        }))),
      }));
    });
  }

  return commands;
};

async function put(commands: CommandsArray): Promise<void> {
  if (!GUILD_ID) {
    console.log('Updating global commands for production bot\n\n');
  } else {
    console.log('Updating guild commands for dev bot\n\n');
  }

  if (commands.length > 100) {
    throw new Error('the maximum number of commands allowed is 100');
  }

  const url = GUILD_ID
    ? `https://discord.com/api/v10/applications/${APP_ID}/guilds/${GUILD_ID}/commands`
    : `https://discord.com/api/v10/applications/${APP_ID}/commands`;

  const response = await fetch(url, {
    method: 'PUT',
    headers: {
      'Authorization': `Bot ${BOT_TOKEN}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(commands),
  });

  if (response.status !== 200) {
    throw new Error(JSON.stringify(await response.json(), undefined, 2));
  } else {
    // console.log(response.status, response.statusText);
    console.log(green('OK'));
  }
}

await put([
  // standard gacha commands
  // uses characters and media from all packs
  ...Command({
    name: 'search',
    aliases: ['anime', 'manga', 'media'],
    description: 'Search for a media',
    options: [
      Option({
        name: 'query',
        description: 'The title of the media',
        autocomplete: true,
        type: Type.STRING,
      }),
      Option({
        name: 'debug',
        description: 'Display the nerdy stuff',
        type: Type.BOOLEAN,
        optional: true,
      }),
    ],
  }),
  ...Command({
    name: 'character',
    description: 'Search for a character',
    options: [
      Option({
        name: 'query',
        description: 'The name of the character',
        autocomplete: true,
        type: Type.STRING,
      }),
      Option({
        name: 'debug',
        description: 'Display the nerdy stuff',
        type: Type.BOOLEAN,
        optional: true,
      }),
    ],
  }),
  ...Command({
    name: 'music',
    description: 'Look up the music and theme songs of a media',
    aliases: ['songs', 'themes'],
    options: [
      Option({
        name: 'query',
        description: 'The title of the media',
        autocomplete: true,
        type: Type.STRING,
      }),
    ],
  }),
  ...Command({
    name: 'now',
    description: 'Check what you can do right now',
    aliases: ['checklist', 'cl', 'tu'],
  }),
  ...Command({
    name: 'gacha',
    description: 'Start a gacha pull',
    aliases: ['pull', 'roll', 'w'],
  }),
  ...Command({
    name: 'collection',
    description: 'View all the characters in your possession',
    aliases: ['list', 'mm'],
  }),
  ...Command({
    name: 'force_pull',
    description: 'Force a gacha pull',
    defaultPermission: Permission.ADMINISTRATORS,
    devOnly: true,
    options: [
      Option({
        name: 'id',
        description: 'The id of the character',
        type: Type.STRING,
      }),
    ],
  }),
  // pack management commands
  ...Command({
    name: 'packs',
    description: 'Pack management commands',
    defaultPermission: Permission.MANAGE_GUILD,
    options: [
      Option({
        name: 'builtin',
        description: 'List all the builtin packs',
        type: Type.SUB_COMMAND,
        optional: true,
      }),
      Option({
        name: 'manual',
        description: 'List all the manually instated packs',
        type: Type.SUB_COMMAND,
        optional: true,
      }),
    ],
  }),
  // non-standard commands (pack commands)
  ...Pack('./packs/anilist'),
]);
