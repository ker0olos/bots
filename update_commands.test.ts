import { assertSnapshot } from 'https://deno.land/std@0.177.0/testing/snapshot.ts';

import { commands } from './update_commands.ts';

Deno.test('commands', async (test) => {
  await assertSnapshot(test, commands);
});
