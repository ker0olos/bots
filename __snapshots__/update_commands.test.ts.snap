export const snapshot = {};

snapshot[`commands 1`] = `
[
  {
    default_member_permissions: undefined,
    description: "Search for a specific series",
    name: "search",
    options: [
      {
        autocomplete: true,
        choices: undefined,
        description: "The title of the media",
        name: "title",
        options: undefined,
        required: true,
        type: 3,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Display the nerdy stuff",
        name: "debug",
        options: undefined,
        required: false,
        type: 5,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "Search for a specific series",
    name: "anime",
    options: [
      {
        autocomplete: true,
        choices: undefined,
        description: "The title of the media",
        name: "title",
        options: undefined,
        required: true,
        type: 3,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Display the nerdy stuff",
        name: "debug",
        options: undefined,
        required: false,
        type: 5,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "Search for a specific series",
    name: "manga",
    options: [
      {
        autocomplete: true,
        choices: undefined,
        description: "The title of the media",
        name: "title",
        options: undefined,
        required: true,
        type: 3,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Display the nerdy stuff",
        name: "debug",
        options: undefined,
        required: false,
        type: 5,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "Search for a specific series",
    name: "media",
    options: [
      {
        autocomplete: true,
        choices: undefined,
        description: "The title of the media",
        name: "title",
        options: undefined,
        required: true,
        type: 3,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Display the nerdy stuff",
        name: "debug",
        options: undefined,
        required: false,
        type: 5,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "Search for a specific character",
    name: "character",
    options: [
      {
        autocomplete: true,
        choices: undefined,
        description: "The name of the character",
        name: "name",
        options: undefined,
        required: true,
        type: 3,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Display the nerdy stuff",
        name: "debug",
        options: undefined,
        required: false,
        type: 5,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "Search for a specific character",
    name: "char",
    options: [
      {
        autocomplete: true,
        choices: undefined,
        description: "The name of the character",
        name: "name",
        options: undefined,
        required: true,
        type: 3,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Display the nerdy stuff",
        name: "debug",
        options: undefined,
        required: false,
        type: 5,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "Search for a specific character",
    name: "im",
    options: [
      {
        autocomplete: true,
        choices: undefined,
        description: "The name of the character",
        name: "name",
        options: undefined,
        required: true,
        type: 3,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Display the nerdy stuff",
        name: "debug",
        options: undefined,
        required: false,
        type: 5,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "New to Fable? We got you",
    name: "help",
    options: undefined,
  },
  {
    default_member_permissions: undefined,
    description: "New to Fable? We got you",
    name: "start",
    options: undefined,
  },
  {
    default_member_permissions: undefined,
    description: "New to Fable? We got you",
    name: "guide",
    options: undefined,
  },
  {
    default_member_permissions: undefined,
    description: "New to Fable? We got you",
    name: "tuto",
    options: undefined,
  },
  {
    default_member_permissions: undefined,
    description: "Check what you can do right now",
    name: "now",
    options: undefined,
  },
  {
    default_member_permissions: undefined,
    description: "Check what you can do right now",
    name: "checklist",
    options: undefined,
  },
  {
    default_member_permissions: undefined,
    description: "Check what you can do right now",
    name: "cl",
    options: undefined,
  },
  {
    default_member_permissions: undefined,
    description: "Check what you can do right now",
    name: "tu",
    options: undefined,
  },
  {
    default_member_permissions: undefined,
    description: "Start a gacha pull",
    name: "gacha",
    options: undefined,
  },
  {
    default_member_permissions: undefined,
    description: "Start a gacha pull",
    name: "w",
    options: undefined,
  },
  {
    default_member_permissions: undefined,
    description: "Start a quiet gacha pull with no animation",
    name: "pull",
    options: undefined,
  },
  {
    default_member_permissions: undefined,
    description: "Start a quiet gacha pull with no animation",
    name: "q",
    options: undefined,
  },
  {
    default_member_permissions: undefined,
    description: "Preform a fake gacha pull (Developer Only)",
    name: "fake_pull",
    options: [
      {
        autocomplete: undefined,
        choices: undefined,
        description: "The id of the character",
        name: "id",
        options: undefined,
        required: true,
        type: 3,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "party management commands",
    name: "party",
    options: [
      {
        autocomplete: undefined,
        choices: undefined,
        description: "View your current party",
        name: "view",
        options: undefined,
        required: false,
        type: 1,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Assign a character to your party",
        name: "assign",
        options: [
          {
            autocomplete: true,
            choices: undefined,
            description: "The name of the character",
            name: "name",
            options: undefined,
            required: true,
            type: 3,
          },
          {
            autocomplete: undefined,
            choices: [
              {
                name: "1",
                value: 1,
              },
              {
                name: "2",
                value: 2,
              },
              {
                name: "3",
                value: 3,
              },
              {
                name: "4",
                value: 4,
              },
              {
                name: "5",
                value: 5,
              },
            ],
            description: "The spot where you want this character",
            name: "spot",
            options: undefined,
            required: false,
            type: 4,
          },
        ],
        required: false,
        type: 1,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Remove a character from your party",
        name: "remove",
        options: [
          {
            autocomplete: undefined,
            choices: [
              {
                name: "1",
                value: 1,
              },
              {
                name: "2",
                value: 2,
              },
              {
                name: "3",
                value: 3,
              },
              {
                name: "4",
                value: 4,
              },
              {
                name: "5",
                value: 5,
              },
            ],
            description: "The spot the character occupies",
            name: "spot",
            options: undefined,
            required: true,
            type: 4,
          },
        ],
        required: false,
        type: 1,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "party management commands",
    name: "team",
    options: [
      {
        autocomplete: undefined,
        choices: undefined,
        description: "View your current party",
        name: "view",
        options: undefined,
        required: false,
        type: 1,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Assign a character to your party",
        name: "assign",
        options: [
          {
            autocomplete: true,
            choices: undefined,
            description: "The name of the character",
            name: "name",
            options: undefined,
            required: true,
            type: 3,
          },
          {
            autocomplete: undefined,
            choices: [
              {
                name: "1",
                value: 1,
              },
              {
                name: "2",
                value: 2,
              },
              {
                name: "3",
                value: 3,
              },
              {
                name: "4",
                value: 4,
              },
              {
                name: "5",
                value: 5,
              },
            ],
            description: "The spot where you want this character",
            name: "spot",
            options: undefined,
            required: false,
            type: 4,
          },
        ],
        required: false,
        type: 1,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Remove a character from your party",
        name: "remove",
        options: [
          {
            autocomplete: undefined,
            choices: [
              {
                name: "1",
                value: 1,
              },
              {
                name: "2",
                value: 2,
              },
              {
                name: "3",
                value: 3,
              },
              {
                name: "4",
                value: 4,
              },
              {
                name: "5",
                value: 5,
              },
            ],
            description: "The spot the character occupies",
            name: "spot",
            options: undefined,
            required: true,
            type: 4,
          },
        ],
        required: false,
        type: 1,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "party management commands",
    name: "p",
    options: [
      {
        autocomplete: undefined,
        choices: undefined,
        description: "View your current party",
        name: "view",
        options: undefined,
        required: false,
        type: 1,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Assign a character to your party",
        name: "assign",
        options: [
          {
            autocomplete: true,
            choices: undefined,
            description: "The name of the character",
            name: "name",
            options: undefined,
            required: true,
            type: 3,
          },
          {
            autocomplete: undefined,
            choices: [
              {
                name: "1",
                value: 1,
              },
              {
                name: "2",
                value: 2,
              },
              {
                name: "3",
                value: 3,
              },
              {
                name: "4",
                value: 4,
              },
              {
                name: "5",
                value: 5,
              },
            ],
            description: "The spot where you want this character",
            name: "spot",
            options: undefined,
            required: false,
            type: 4,
          },
        ],
        required: false,
        type: 1,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Remove a character from your party",
        name: "remove",
        options: [
          {
            autocomplete: undefined,
            choices: [
              {
                name: "1",
                value: 1,
              },
              {
                name: "2",
                value: 2,
              },
              {
                name: "3",
                value: 3,
              },
              {
                name: "4",
                value: 4,
              },
              {
                name: "5",
                value: 5,
              },
            ],
            description: "The spot the character occupies",
            name: "spot",
            options: undefined,
            required: true,
            type: 4,
          },
        ],
        required: false,
        type: 1,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "collection browsing commands",
    name: "collection",
    options: [
      {
        autocomplete: undefined,
        choices: undefined,
        description: "View all your stars",
        name: "stars",
        options: [
          {
            autocomplete: undefined,
            choices: [
              {
                name: "5",
                value: 5,
              },
              {
                name: "4",
                value: 4,
              },
              {
                name: "3",
                value: 3,
              },
              {
                name: "2",
                value: 2,
              },
              {
                name: "1",
                value: 1,
              },
            ],
            description: "The star rating",
            name: "rating",
            options: undefined,
            required: true,
            type: 4,
          },
          {
            autocomplete: undefined,
            choices: undefined,
            description: "List someone else's stars",
            name: "user",
            options: undefined,
            required: false,
            type: 6,
          },
        ],
        required: false,
        type: 1,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "View your characters from a specific media",
        name: "media",
        options: [
          {
            autocomplete: true,
            choices: undefined,
            description: "The title of the media",
            name: "title",
            options: undefined,
            required: true,
            type: 3,
          },
          {
            autocomplete: undefined,
            choices: undefined,
            description: "View someone else's stars",
            name: "user",
            options: undefined,
            required: false,
            type: 6,
          },
        ],
        required: false,
        type: 1,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "collection browsing commands",
    name: "coll",
    options: [
      {
        autocomplete: undefined,
        choices: undefined,
        description: "View all your stars",
        name: "stars",
        options: [
          {
            autocomplete: undefined,
            choices: [
              {
                name: "5",
                value: 5,
              },
              {
                name: "4",
                value: 4,
              },
              {
                name: "3",
                value: 3,
              },
              {
                name: "2",
                value: 2,
              },
              {
                name: "1",
                value: 1,
              },
            ],
            description: "The star rating",
            name: "rating",
            options: undefined,
            required: true,
            type: 4,
          },
          {
            autocomplete: undefined,
            choices: undefined,
            description: "List someone else's stars",
            name: "user",
            options: undefined,
            required: false,
            type: 6,
          },
        ],
        required: false,
        type: 1,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "View your characters from a specific media",
        name: "media",
        options: [
          {
            autocomplete: true,
            choices: undefined,
            description: "The title of the media",
            name: "title",
            options: undefined,
            required: true,
            type: 3,
          },
          {
            autocomplete: undefined,
            choices: undefined,
            description: "View someone else's stars",
            name: "user",
            options: undefined,
            required: false,
            type: 6,
          },
        ],
        required: false,
        type: 1,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "collection browsing commands",
    name: "mm",
    options: [
      {
        autocomplete: undefined,
        choices: undefined,
        description: "View all your stars",
        name: "stars",
        options: [
          {
            autocomplete: undefined,
            choices: [
              {
                name: "5",
                value: 5,
              },
              {
                name: "4",
                value: 4,
              },
              {
                name: "3",
                value: 3,
              },
              {
                name: "2",
                value: 2,
              },
              {
                name: "1",
                value: 1,
              },
            ],
            description: "The star rating",
            name: "rating",
            options: undefined,
            required: true,
            type: 4,
          },
          {
            autocomplete: undefined,
            choices: undefined,
            description: "List someone else's stars",
            name: "user",
            options: undefined,
            required: false,
            type: 6,
          },
        ],
        required: false,
        type: 1,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "View your characters from a specific media",
        name: "media",
        options: [
          {
            autocomplete: true,
            choices: undefined,
            description: "The title of the media",
            name: "title",
            options: undefined,
            required: true,
            type: 3,
          },
          {
            autocomplete: undefined,
            choices: undefined,
            description: "View someone else's stars",
            name: "user",
            options: undefined,
            required: false,
            type: 6,
          },
        ],
        required: false,
        type: 1,
      },
    ],
  },
  {
    default_member_permissions: 32,
    description: "pack management commands",
    name: "packs",
    options: [
      {
        autocomplete: undefined,
        choices: undefined,
        description: "List all the builtin packs",
        name: "builtin",
        options: undefined,
        required: false,
        type: 1,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "List all instated community packs",
        name: "community",
        options: undefined,
        required: false,
        type: 1,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Install a community pack",
        name: "install",
        options: [
          {
            autocomplete: undefined,
            choices: undefined,
            description: "A github repository url",
            name: "github",
            options: undefined,
            required: true,
            type: 3,
          },
          {
            autocomplete: undefined,
            choices: undefined,
            description: "A ref to a branch or commit sha",
            name: "ref",
            options: undefined,
            required: false,
            type: 3,
          },
        ],
        required: false,
        type: 1,
      },
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Validate a community pack's manifest",
        name: "validate",
        options: [
          {
            autocomplete: undefined,
            choices: undefined,
            description: "A github repository url",
            name: "github",
            options: undefined,
            required: true,
            type: 3,
          },
          {
            autocomplete: undefined,
            choices: undefined,
            description: "A ref to a branch or commit sha",
            name: "ref",
            options: undefined,
            required: false,
            type: 3,
          },
        ],
        required: false,
        type: 1,
      },
    ],
  },
  {
    default_member_permissions: undefined,
    description: "A pack powered by [AniList](https://anilist.co).

Contains over 100K of anime and manga characters.",
    name: "anilist",
    options: [
      {
        autocomplete: undefined,
        choices: undefined,
        description: "Find when the next episode of an anime is airing",
        name: "next_episode",
        options: [
          {
            autocomplete: undefined,
            choices: undefined,
            description: "The title for an anime",
            name: "title",
            options: undefined,
            required: true,
            type: 3,
          },
        ],
        required: false,
        type: 1,
      },
    ],
  },
]
`;
